### DISCUSSION TITLE
Error trying to apply DeepExplainer on a neural network

### DISCUSSION THREAD

**Discussion by NDincic** [shap/shap#1533] [Q&A]
Dear community,

I am quite new to coding in Python so please excuse any lack of expertise or incomplete information I might be giving. I was hoping that some of you here could maybe help me:

I am doing a research project on explaining an already existing neural network which was created by a fellow student earlier. For the sake of data protection and simplicity, I adapted the code to the (well-known) titanic dataset. I am trying to apply SHAP and (to the best of my knowledge) mastered the basic math and code behind the application. For excercise, I already applied the TreeExplainer to a random forrest and played with the plotting possiblities in Jupyter Notebook. Now I am trying to apply the DeepExplainer to my code and simply can't sort out the following error:

____________________________________________________________________________

Traceback (most recent call last):

  File "C:\Users\n.dincic\Documents\Python Scripts\Studienarbeit\Übung\Titanic\Titanic.py", line 195, in <module>
    explainer = shap.DeepExplainer(model, data=training_x)

  File "C:\Users\n.dincic\Anaconda3\lib\site-packages\shap\explainers\deep\__init__.py", line 82, in __init__
    self.explainer = PyTorchDeepExplainer(model, data)

  File "C:\Users\n.dincic\Anaconda3\lib\site-packages\shap\explainers\deep\deep_pytorch.py", line 54, in __init__
    outputs = model(*data)

  File "C:\Users\n.dincic\Anaconda3\lib\site-packages\torch\nn\modules\module.py", line 722, in _call_impl
    result = self.forward(*input, **kwargs)

  File "C:\Users\n.dincic\Documents\Python Scripts\Studienarbeit\Übung\Titanic\Titanic.py", line 83, in forward
    x = x.view(x.shape[0], -1)

ValueError: Type must be a sub-type of ndarray type
_____________________________________________________________________________

I am running the following code:

_____________________________________________________________________________
```
import pandas as pd
training_data_path = "C:\\Users\\n.dincic\\Documents\\Python Scripts\\Studienarbeit\\Übung\\Titanic\\train_data.csv"
testing_data_path = "C:\\Users\\n.dincic\\Documents\\Python Scripts\\Studienarbeit\\Übung\\Titanic\\test_data.csv"
training = pd.read_csv(training_data_path)   
testing_data = pd.read_csv(testing_data_path)
training_x = training.drop(['Survived' , 'PassengerId'],axis='columns').values
training_y = training['Survived'].values
[Titanic.zip](https://github.com/slundberg/shap/files/5403419/Titanic.zip)

testing_x = testing_data.drop(['Survived' , 'PassengerId'],axis='columns').values
testing_y = testing_data['Survived'].values

C=3
gamma=0.009
classifier=svm.SVC(C=C,gamma=gamma, probability=True)
classifier.fit(training_x,training_y)

import numpy as np
import pandas as pd
from torch import nn, optim
import torch.nn.functional as F
import torch.utils.data as data
import torch

class TitanicDataset(data.Dataset):
    def __init__(self):
        train = training_x
        train_labels = training_y
        self.datalist = train
        self.labels = train_labels
    def __getitem__(self, index):
        return torch.Tensor(self.datalist[index].astype(float)), self.labels[index]
    def __len__(self):
        return self.datalist.shape[0]
    
train_Set = TitanicDataset()
trainloader = torch.utils.data.DataLoader( dataset = train_Set , batch_size= 8 , shuffle = True) 

class Classifier(nn.Module):
    def __init__(self):
        super().__init__() 
        self.fc1 = nn.Linear(15, 12)
        self.fc2 = nn.Linear(12, 4)
        self.fc3 = nn.Linear(4, 2)
        
        self.dropout = nn.Dropout(p=0.01)
        
    def forward(self, x):
        x = x.view(x.shape[0], -1)

        x = torch.sigmoid(self.fc1(x)) 
        x = self.dropout(torch.sigmoid(self.fc2(x)))
        x = F.log_softmax(self.fc3(x), dim=1) 

        return x

model = Classifier()
criterion = nn.NLLLoss()
optimizer =  optim.SGD(model.parameters(), lr=0.01, momentum=0.9)

anzahl_trainingsepochen = int(30)
from torch.autograd import Variable
for epoch in range(anzahl_trainingsepochen):
    for i, (Titanic, labels) in enumerate(trainloader):
        Titanic = Variable(Titanic)
        labels = Variable(labels)
        
        
        optimizer.zero_grad()
        outputs = model(Titanic)
        
        loss = criterion(outputs, labels)
        loss.backward()
        if (i+1) % 100 == 0:
            print ('Epoch [%d/%d], Iter [%d] Loss: %.4f' %(epoch+1, anzahl_trainingsepochen, i+1, loss.data))

model.eval()
test = testing_x
test = Variable(torch.Tensor(test))

pred = model(test)

_, predlabel = torch.max(pred.data, 1)
predlabel = predlabel.tolist()

predlabel = pd.DataFrame(predlabel)
predlabel.index = np.arange(100) + 1 
id = np.arange(100) + 1
id = pd.DataFrame(id)
id.index = id.index + 1

predlabel = pd.concat([id,predlabel], axis=1)
predlabel.columns = ["No.", "Survived"]

predlabel.to_csv('predict.csv', index= False)

import shap

explainer = shap.DeepExplainer(model, data=training_x)
shap_values = explainer.shap_values(testing_x)
```
______________________________________________________________________________

If it is helpful in any way, I attached the Input data in a zip-file. It is just the pre-processed titanic dataset.

I use the predict.csv to evaluate the performance (which is poor, but this is not my main problem) in excel afterwards.

As far as I understand, x is the inner variable (or matrix) of the model that gets passed on from layer to layer. The shap algorithm tries to read x but datatype doesn't seem to be right. Did I get that the right way? Do you think there is a way to fix it, or is the code architecture itself simply unsuitable for performing shap? If my question is completely out of line, you may also tell me. I am thankful for any input. 

Hope I didn't make any mistakes while copying the code, I will try to answer any questions arising. Thank you very much for your precious time!
Nikola

**Comment by jaxondk** [comment]
Hi Nikola,

It looks like you just have a type error that is being thrown within `model.forward`. The stack trace shows you that when you call `explainer = shap.DeepExplainer(model, data=training_x)`, eventually your model's forward function gets called and  `x = x.view(x.shape[0], -1)` is throwing an error because x needs to be a sub type of the numpy ndarray class. 

So, you should investigate to see what type `training_x` is when you pass it into DeepExplainer, and then see what type it is when model.forward is called. You can do this in python with `type(obj)`, where obj would be `training_x` when you're checking the type before passing it to DeepExplainer, and just `x` when you're in model.forward. If `type(x)` in forward isn't a sub type of ndarray but `type(training_x)` is, then the type of training_x is getting changed somewhere.

As for your more conceptual questions, the `forward` function of a pytorch module is indeed what passes information from module to module. So x is always the input into the current module. However, a pytorch module is often itself made up of many layers, and often represents an entire model (like yours does here). So, x is simply the input to your model, and what forward should return is the model's predictions. In your forward function, you pass x through each FC layer, and then return the softmaxed probabilities. 

And no, this type of model will work fine with SHAP. It's likely that you just have an error, that hopefully you can find by looking into the type of x when it gets passed into forward.

Hope it helps!
Jaxon

PS
You may want to reformat your post to put your code in a code block so its easier for the community to read through your post and help you out. This can be done with three backticks in a row before and after your code block, like this:

```
It makes reading your code much easier
   Especially code indentation
```
